File: scripts/fetch-zh.mjs
/**
 * Build the Chinese localisation snapshot.
 *
 * Sources
 *   • Player names  — Wikidata (WTA player id = P597), normalised to Simplified
 *                     Chinese via the MediaWiki variant converter.
 *   • Tournaments   — curated dictionary for the calendar, plus Wikidata for
 *                     anything the dictionary misses.
 *   • Countries, rounds, surfaces, levels — curated terminology.
 *
 * Produces `data/zh.json`, which the front end merges with the English snapshots
 * so every player, event, surface and round can be shown bilingually.
 */
import { readFile } from 'node:fs/promises';
import { resolve } from 'node:path';
import { ROOT, log } from './lib.mjs';
import {
  sparql,
  toSimplified,
  writeJsonFile,
  countryZh,
  tournamentZh,
  LEVEL_ZH,
  ROUND_ZH,
  SURFACE_ZH,
  COUNTRY_ZH,
  VARIANT_CACHE,
  PLAYER_ZH,
  playerZhFallback,
} from './zh-terms.mjs';

const read = async (f, fallback) => {
  try {
    return JSON.parse(await readFile(resolve(ROOT, f), 'utf8'));
  } catch {
    return fallback;
  }
};

log('zh', 'Building Chinese localisation snapshot…');

const rankings = await read('data/rankings-singles.json', { players: [] });
const bios = await read('data/bios.json', {});
const matches = await read('data/matches.json', {});
const tournaments = await read('data/tournaments.json', { events: [] });
const h2hNames = await read('data/h2h-names.json', {});

/* ------------------------------------------------------------------ */
/* 1. Every player id we need a Chinese name for                       */
/* ------------------------------------------------------------------ */

const wanted = new Map(); // wtaId(string) -> english name
for (const p of rankings.players) wanted.set(String(p.id), p.name);
for (const [id, b] of Object.entries(bios)) if (!wanted.has(String(id))) wanted.set(String(id), b.name);
for (const [id, n] of Object.entries(h2hNames)) if (!wanted.has(String(id))) wanted.set(String(id), n.n);
for (const list of Object.values(matches)) {
  for (const m of list) if (m.oid && m.o) wanted.set(String(m.oid), m.o);
}
for (const e of tournaments.events) {
  if (e.champion?.id) wanted.set(String(e.champion.id), e.champion.name);
  for (const d of e.doublesChampions || []) if (d.id) wanted.set(String(d.id), d.name);
}

log('zh', `  ${wanted.size} distinct players across rankings, biographies, match logs and champions`);

/* ------------------------------------------------------------------ */
/* 2. Wikidata lookup by WTA id                                        */
/* ------------------------------------------------------------------ */

log('zh', '  Querying Wikidata for WTA ids and Chinese labels…');
let rows = [];
try {
  rows = await sparql(`SELECT ?wtaId ?zh ?zhHans ?zhCn ?zhSg ?en WHERE {
    ?p wdt:P597 ?wtaId .
    OPTIONAL { ?p rdfs:label ?zh . FILTER(lang(?zh)="zh") }
    OPTIONAL { ?p rdfs:label ?zhHans . FILTER(lang(?zhHans)="zh-hans") }
    OPTIONAL { ?p rdfs:label ?zhCn . FILTER(lang(?zhCn)="zh-cn") }
    OPTIONAL { ?p rdfs:label ?zhSg . FILTER(lang(?zhSg)="zh-sg") }
    OPTIONAL { ?p rdfs:label ?en . FILTER(lang(?en)="en") }
  }`);
} catch (err) {
  log('zh', `  ! Wikidata query failed: ${err.message}`);
}

const byWtaId = new Map();
for (const r of rows) {
  const id = r.wtaId?.value;
  if (!id) continue;
  const label =
    r.zhHans?.value || r.zhCn?.value || r.zhSg?.value || r.zh?.value || '';
  const existing = byWtaId.get(id);
  // Prefer an entry that actually carries a Chinese label.
  if (!existing || (!existing.zh && label)) {
    byWtaId.set(id, { zh: label, en: r.en?.value || '' });
  }
}

log('zh', `  Wikidata returned ${byWtaId.size} WTA-indexed players`);

/* ------------------------------------------------------------------ */
/* 3. Normalise everything to Simplified Chinese                       */
/* ------------------------------------------------------------------ */

const rawZh = [];
const playerZh = {};

for (const [id, name] of wanted) {
  const wd = byWtaId.get(id);
  if (wd?.zh) {
    rawZh.push(wd.zh);
    playerZh[id] = wd.zh; // converted below
  }
}

// Tournament names: curated dictionary first, Wikidata/city fallback after.
const tournamentZhMap = {};
const eventNamesRaw = new Set();
for (const e of tournaments.events) {
  const curated = tournamentZh(e.name, e.city);
  if (curated) {
    tournamentZhMap[e.name] = curated;
  } else if (e.title) {
    eventNamesRaw.add(e.name);
  }
}

// Match-log event names (ITF and lower-tier events are frequent here).
const matchEventNames = new Set();
for (const list of Object.values(matches)) {
  for (const m of list) {
    if (!tournamentZhMap[m.t]) matchEventNames.add(m.t);
  }
}

log('zh', `  Translating ${eventNamesRaw.size} calendar events and ${matchEventNames.size} match-log events…`);

// Ask Wikidata for Chinese labels of tennis tournaments matching these names.
async function wikidataTournaments(names) {
  const list = [...names];
  const out = new Map();
  const CHUNK = 160;
  for (let i = 0; i < list.length; i += CHUNK) {
    const chunk = list
      .slice(i, i + CHUNK)
      .map((n) => `"${n.replace(/["\\]/g, '')}"@en`)
      .join(' ');
    try {
      const rows = await sparql(`SELECT ?name ?labelZh WHERE {
        VALUES ?name { ${chunk} }
        ?item rdfs:label ?name .
        ?item rdfs:label ?labelZh . FILTER(lang(?labelZh)="zh")
      }`);
      for (const r of rows) {
        const n = r.name?.value;
        const z = r.labelZh?.value;
        if (n && z && !out.has(n)) out.set(n, z);
      }
    } catch (err) {
      log('zh', `  ! tournament label batch failed: ${err.message}`);
    }
  }
  return out;
}

try {
  const wdTournaments = await wikidataTournaments([...eventNamesRaw, ...matchEventNames]);
  log('zh', `  Wikidata resolved ${wdTournaments.size} tournament labels`);
  for (const [name, label] of wdTournaments) {
    rawZh.push(label);
    tournamentZhMap[name] = label;
  }
} catch (err) {
  log('zh', `  ! tournament lookup skipped: ${err.message}`);
}

/* ------------------------------------------------------------------ */
/* 4. Convert everything to Simplified Chinese in batches              */
/* ------------------------------------------------------------------ */

const uniqueZh = [...new Set(rawZh)];
log('zh', `  Converting ${uniqueZh.length} names to Simplified Chinese…`);
const converted = await toSimplified(uniqueZh, { cacheFile: VARIANT_CACHE });

const toHans = (s) => (s && converted.has(s) ? converted.get(s) : s);

const playerZhFinal = {};
for (const [id, z] of Object.entries(playerZh)) playerZhFinal[id] = toHans(z);

// Fill the gaps from the curated transliteration list.
let curatedFilled = 0;
for (const [id, name] of wanted) {
  if (playerZhFinal[id]) continue;
  const fb = playerZhFallback(name);
  if (fb) {
    playerZhFinal[id] = fb;
    curatedFilled += 1;
  }
}
for (const [name, z] of Object.entries(tournamentZhMap)) tournamentZhMap[name] = toHans(z);

/* ------------------------------------------------------------------ */
/* 5. Fill tournament gaps with a curated city-based fallback          */
/* ------------------------------------------------------------------ */

let fallbackCount = 0;
for (const name of new Set([...eventNamesRaw, ...matchEventNames])) {
  if (tournamentZhMap[name]) continue;
  const fb = tournamentZh(name, '');
  tournamentZhMap[name] = fb || '';
  if (fb) fallbackCount += 1;
}

/* ------------------------------------------------------------------ */
/* 6. Snapshot                                                            */
/* ------------------------------------------------------------------ */

const events = Object.values(tournamentZhMap).filter(Boolean).length;

const payload = {
  generatedAt: new Date().toISOString(),
  sources: {
    playerNames: 'Wikidata (property P597 — WTA player id)',
    simplified: 'MediaWiki zh-hans variant conversion',
    terminology: 'Curated tennis terminology',
  },
  players: playerZhFinal,
  tournaments: tournamentZhMap,
  countries: COUNTRY_ZH,
  levels: LEVEL_ZH,
  rounds: ROUND_ZH,
  surfaces: SURFACE_ZH,
};

await writeJsonFile('data/zh.json', payload);

log(
  'zh',
  `Done — ${Object.keys(playerZhFinal).length} player names (${curatedFilled} via curated transliteration), ` +
    `${events} tournament names (${fallbackCount} via city fallback), ${Object.keys(COUNTRY_ZH).length} country names.`,
);

// Report coverage against the current ranking snapshot so gaps are visible.
const ranked = rankings.players;
const covered = ranked.filter((p) => playerZhFinal[String(p.id)]).length;
log('zh', `Ranking coverage: ${covered}/${ranked.length} players have a Chinese name.`);
const missing = ranked.filter((p) => !playerZhFinal[String(p.id)]).slice(0, 12);
if (missing.length) {
  log('zh', `  no Chinese name: ${missing.map((p) => `${p.rank} ${p.name}`).join(', ')}${missing.length === 12 ? ' …' : ''}`);
}

File: pyscripts/fetch_zh.py
#!/usr/bin/env python3
"""
Build the Chinese localisation snapshot.

Sources
  • Player names  — Wikidata (WTA player id = P597), normalised to Simplified
                    Chinese via the MediaWiki variant converter.
  • Tournaments   — curated dictionary for the calendar, plus Wikidata for
                    anything the dictionary misses.
  • Countries, rounds, surfaces, levels — curated terminology.

Produces `data/zh.json`, which the site generator merges with the English
snapshots so every player, event, surface and round can be shown bilingually.
"""

from __future__ import annotations

from datetime import datetime, timezone

from wtalib import log, read_json, write_json
from zh_terms import (
    COUNTRY_ZH,
    LEVEL_ZH,
    PLAYER_ZH,
    ROUND_ZH,
    SURFACE_ZH,
    VARIANT_CACHE,
    has_cjk,
    player_zh_fallback,
    sparql,
    to_simplified,
    tournament_zh,
)


def collect_wanted(rankings, bios, matches, tournaments, h2h_names) -> dict:
    """Every player id that needs a Chinese name, mapped to its English name."""
    wanted: dict[str, str] = {}
    for player in rankings.get("players", []):
        wanted[str(player["id"])] = player["name"]
    for pid, bio in bios.items():
        wanted.setdefault(str(pid), bio.get("name", ""))
    for pid, entry in h2h_names.items():
        wanted.setdefault(str(pid), entry.get("n", ""))
    for rows in matches.values():
        for match in rows:
            if match.get("oid") and match.get("o"):
                wanted.setdefault(str(match["oid"]), match["o"])
    for event in tournaments.get("events", []):
        champion = event.get("champion")
        if champion and champion.get("id"):
            wanted.setdefault(str(champion["id"]), champion.get("name", ""))
        for pair in event.get("doublesChampions") or []:
            if pair.get("id"):
                wanted.setdefault(str(pair["id"]), pair.get("name", ""))
    return wanted


def fetch_wikidata_players() -> dict:
    """WTA-id-indexed Chinese labels from Wikidata."""
    query = """SELECT ?wtaId ?zh ?zhHans ?zhCn ?zhSg ?en WHERE {
      ?p wdt:P597 ?wtaId .
      OPTIONAL { ?p rdfs:label ?zh . FILTER(lang(?zh)="zh") }
      OPTIONAL { ?p rdfs:label ?zhHans . FILTER(lang(?zhHans)="zh-hans") }
      OPTIONAL { ?p rdfs:label ?zhCn . FILTER(lang(?zhCn)="zh-cn") }
      OPTIONAL { ?p rdfs:label ?zhSg . FILTER(lang(?zhSg)="zh-sg") }
      OPTIONAL { ?p rdfs:label ?en . FILTER(lang(?en)="en") }
    }"""
    by_id: dict[str, dict] = {}
    for row in sparql(query):
        wta_id = (row.get("wtaId") or {}).get("value")
        if not wta_id:
            continue
        label = (
            (row.get("zhHans") or {}).get("value")
            or (row.get("zhCn") or {}).get("value")
            or (row.get("zhSg") or {}).get("value")
            or (row.get("zh") or {}).get("value")
            or ""
        )
        existing = by_id.get(wta_id)
        if not existing or (not existing["zh"] and label):
            by_id[wta_id] = {"zh": label, "en": (row.get("en") or {}).get("value", "")}
    return by_id


def fetch_wikidata_tournaments(names: list[str]) -> dict:
    """Chinese labels for tournament names, queried in chunks."""
    out: dict[str, str] = {}
    chunk_size = 160
    for start in range(0, len(names), chunk_size):
        chunk = names[start:start + chunk_size]
        values = " ".join('"' + n.replace('"', "").replace("\\", "") + '"@en' for n in chunk)
        query = f"""SELECT ?name ?labelZh WHERE {{
          VALUES ?name {{ {values} }}
          ?item rdfs:label ?name .
          ?item rdfs:label ?labelZh . FILTER(lang(?labelZh)="zh")
        }}"""
        try:
            for row in sparql(query):
                name = (row.get("name") or {}).get("value")
                label = (row.get("labelZh") or {}).get("value")
                if name and label:
                    out.setdefault(name, label)
        except RuntimeError as err:
            log("zh", f"  ! tournament label batch failed: {err}")
    return out


def main() -> int:
    log("zh", "Building Chinese localisation snapshot…")

    rankings = read_json("rankings-singles.json", {"players": []})
    bios = read_json("bios.json", {})
    matches = read_json("matches.json", {})
    tournaments = read_json("tournaments.json", {"events": []})
    h2h_names = read_json("h2h-names.json", {})

    wanted = collect_wanted(rankings, bios, matches, tournaments, h2h_names)
    log("zh", f"  {len(wanted)} distinct players across rankings, biographies, match logs and champions")

    log("zh", "  Querying Wikidata for WTA ids and Chinese labels…")
    by_wta_id = {}
    try:
        by_wta_id = fetch_wikidata_players()
    except RuntimeError as err:
        log("zh", f"  ! Wikidata query failed: {err}")
    log("zh", f"  Wikidata returned {len(by_wta_id)} WTA-indexed players")

    # ---------------------------------------------------------------- names
    raw_zh: list[str] = []
    player_zh: dict[str, str] = {}
    for pid, name in wanted.items():
        entry = by_wta_id.get(pid)
        if entry and entry.get("zh"):
            raw_zh.append(entry["zh"])
            player_zh[pid] = entry["zh"]

    # ---------------------------------------------------------- tournaments
    tournament_map: dict[str, str] = {}
    unmatched_events: set[str] = set()
    for event in tournaments.get("events", []):
        curated = tournament_zh(event.get("name"), event.get("city"))
        if curated:
            tournament_map[event["name"]] = curated
        elif event.get("title"):
            unmatched_events.add(event["name"])

    match_events: set[str] = set()
    for rows in matches.values():
        for match in rows:
            if match.get("t") and match["t"] not in tournament_map:
                match_events.add(match["t"])

    log(
        "zh",
        f"  Translating {len(unmatched_events)} calendar events and {len(match_events)} match-log events…",
    )

    try:
        wd_tournaments = fetch_wikidata_tournaments(sorted(unmatched_events | match_events))
        log("zh", f"  Wikidata resolved {len(wd_tournaments)} tournament labels")
        for name, label in wd_tournaments.items():
            raw_zh.append(label)
            tournament_map[name] = label
    except RuntimeError as err:
        log("zh", f"  ! tournament lookup skipped: {err}")

    # ------------------------------------------------- simplified conversion
    unique_zh = list(dict.fromkeys(raw_zh))
    log("zh", f"  Converting {len(unique_zh)} names to Simplified Chinese…")
    converted = to_simplified(unique_zh, cache_file=VARIANT_CACHE)

    def to_hans(value: str) -> str:
        return converted.get(value, value) if value else value

    player_final: dict[str, str] = {}
    latin_only = 0
    for pid, value in player_zh.items():
        hans = to_hans(value)
        # Wikidata sometimes carries a "Chinese" label that is simply the Latin
        # name; treating that as a translation would print the name twice.
        if not has_cjk(hans):
            latin_only += 1
            continue
        player_final[pid] = hans

    tournament_final = {name: to_hans(value) for name, value in tournament_map.items()}

    # ------------------------------------------------- curated transliterations
    curated = 0
    for pid, name in wanted.items():
        if pid in player_final:
            continue
        fallback = player_zh_fallback(name)
        if fallback:
            player_final[pid] = fallback
            curated += 1

    if latin_only:
        log("zh", f'  discarded {latin_only} Latin-only "Chinese" label(s) from Wikidata')

    # ------------------------------------------------------ city-name fallback
    fallback_count = 0
    for name in unmatched_events | match_events:
        if tournament_final.get(name):
            continue
        fallback = tournament_zh(name, "")
        tournament_final[name] = fallback or ""
        if fallback:
            fallback_count += 1

    payload = {
        "generatedAt": datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z"),
        "sources": {
            "playerNames": "Wikidata (property P597 — WTA player id)",
            "simplified": "MediaWiki zh-hans variant conversion",
            "terminology": "Curated tennis terminology",
        },
        "players": player_final,
        "tournaments": tournament_final,
        "countries": COUNTRY_ZH,
        "levels": LEVEL_ZH,
        "rounds": ROUND_ZH,
        "surfaces": SURFACE_ZH,
    }
    write_json("data/zh.json", payload)

    events = sum(1 for v in tournament_final.values() if v)
    log(
        "zh",
        f"Done — {len(player_final)} player names ({curated} via curated transliteration), "
        f"{events} tournament names ({fallback_count} via city fallback), "
        f"{len(COUNTRY_ZH)} country names.",
    )

    covered = sum(1 for p in rankings["players"] if str(p["id"]) in player_final)
    log("zh", f"Ranking coverage: {covered}/{len(rankings['players'])} players have a Chinese name.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
